src/EthersSafe.ts

```diff
@@ -1,28 +1,65 @@
-import { BigNumber, Wallet } from 'ethers'
+import { Provider } from '@ethersproject/providers'
+import { BigNumber, ContractTransaction, Wallet } from 'ethers'
 import { GnosisSafe } from '../typechain'
 import SafeAbi from './abis/SafeAbiV1-2-0.json'
 import Safe from './Safe'
 import { areAddressesEqual } from './utils'
-import { EthSignSignature, SafeSignature } from './utils/signatures'
+import { generatePreValidatedSignature } from './utils/signatures'
+import { EthSignSignature, SafeSignature } from './utils/signatures/SafeSignature'
 import { SafeTransaction } from './utils/transactions'
 
 class EthersSafe implements Safe {
-  #contract: GnosisSafe
+  #contract!: GnosisSafe
   #ethers: any
-  #signer: Wallet
+  #provider!: Provider
+  #signer?: Wallet
 
   /**
    * Creates an instance of the Safe Core SDK.
    *
    * @param ethers - Ethers v5 library
-   * @param signer - Ethers signer
    * @param safeAddress - The address of the Safe account to use
+   * @param providerOrSigner - Ethers provider or signer
    * @returns The Safe Core SDK instance
    */
-  constructor(ethers: any, signer: Wallet, safeAddress: string) {
+  constructor(ethers: any, safeAddress: string, providerOrSigner: Provider | Wallet) {
     this.#ethers = ethers
-    this.#signer = signer
-    this.#contract = new ethers.Contract(safeAddress, SafeAbi, signer)
+    this.connect(safeAddress, providerOrSigner)
+  }
+
+  /**
+   * Initializes the Safe Core SDK connecting the providerOrSigner to the safeAddress.
+   *
+   * @param safeAddress - The address of the Safe account to use
+   * @param providerOrSigner - Ethers provider or signer
+   */
+  connect(safeAddress: string, providerOrSigner: Provider | Wallet): void {
```

Comment:
Would it make more sense to return a new instance (I think ethers normally works like that if you connect a new signer or provider to an existing instance

Reply:
Are you talking about having something like this?

First common block (wallets initialization):
```js
const provider = ethers.getDefaultProvider('homestead')
const wallet1 = ethers.Wallet.createRandom().connect(provider)
const wallet2 = ethers.Wallet.createRandom().connect(provider)
const wallet3 = ethers.Wallet.createRandom().connect(provider)
```
Option 1:
```js
const safeSdk1 = new EthersSafe(ethers, safeAddress, wallet1) //signer = wallet1
const safeSdk2 = safeSdk1.connect(safeAddress, wallet2) //signer = wallet2
const safeSdk3 = safeSdk1.connect(safeAddress, wallet3) //signer = wallet3
```
Option 2:
```js
const safeSdk = new EthersSafe(ethers, safeAddress) // uses default ethers provider, no signer
safeSdk1 = safeSdk.connect(wallet1) //signer = wallet1
safeSdk2 = safeSdk.connect(wallet2) //signer = wallet2
safeSdk3 = safeSdk.connect(wallet3) //signer = wallet3
```
Last common block (signatures and execution):
```js
const tx = { ... }
await safeSdk1.signTransaction(tx) // signed by wallet1
await safeSdk2.signTransaction(tx) // signed by wallet2
await safeSdk3.executeTransaction(tx) // signed and executed by wallet3
```